rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            WT_EXECUTEINWAITTHREAD,
+        },
+    };
+
+    /// Opens and returns the IPv4 and IPv6 registry keys
+    ///
+    /// `HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Services/Tcpip[6]/Parameters/Interfaces`
+    fn open_network_registry_keys() -> Result<[winreg::RegKey; 2]> {
+        let hklm = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+        let path_4 = std::path::Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip")
+            .join("Parameters")
+            .join("Interfaces");
+        let path_6 = std::path::Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip6")
+            .join("Parameters")
+            .join("Interfaces");
+        let flags = winreg::enums::KEY_NOTIFY;
+        Ok([
+            hklm.open_subkey_with_flags(path_4, flags)?,
+            hklm.open_subkey_with_flags(path_6, flags)?,
+        ])
+    }
+
+    pub(crate) fn run_debug() -> Result<()> {
+        tracing_subscriber::fmt::init();
+        let rt = tokio::runtime::Runtime::new()?;
+
+        let [key_ipv4, key_ipv6] = open_network_registry_keys()?;
+
+        let mut listener_4 = Listener::new(key_ipv4)?;
+        let mut listener_6 = Listener::new(key_ipv6)?;
+
+        rt.block_on(async move {
+            loop {
+                tokio::select! {
+                    _r = tokio::signal::ctrl_c() => break,
+                    r = listener_4.notified() => r?,
+                    r = listener_6.notified() => r?,
+                }
+
+                let resolvers = crate::client::resolvers::get()?;
+                tracing::info!(?resolvers);
+            }
+
+            listener_4.close()?;
+            listener_6.close()?;
+            Ok::<_, anyhow::Error>(())
+        })?;
+
+        Ok(())
+    }
+
+    pub(crate) struct CombinedListener {
+        listener_4: Listener,
+        listener_6: Listener,
+    }
+
+    impl CombinedListener {
+        pub(crate) fn new() -> Result<Self> {
+            let [key_ipv4, key_ipv6] = open_network_registry_keys()?;
+            let listener_4 = Listener::new(key_ipv4)?;
+            let listener_6 = Listener::new(key_ipv6)?;
+
+            Ok(Self {
+                listener_4,
+                listener_6,
+            })
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            self.listener_4.close()?;
+            self.listener_6.close()?;
+            tracing::debug!("Unregistered registry listener");
+            Ok(())
+        }
+
+        pub(crate) async fn notified(&mut self) -> Result<()> {
+            tokio::select! {
+                r = self.listener_4.notified() => r?,
+                r = self.listener_6.notified() => r?,
+            }
+            Ok(())
+        }
+    }
+
+    struct Listener {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
```

Comment:
I added a comment to explain it, since it's safety-related and I should have it written down: https://github.com/firezone/firezone/pull/4198/files#diff-f8b44e003aaf67d1e8ce86b898c00ae49ebb1ba2ac6d5ae40ec47fdcf64e996dR504-R507

Reply:
this applies here too https://github.com/firezone/firezone/pull/4198#discussion_r1533135656